kbfsgit/runner.go

```diff
@@ -579,62 +596,127 @@ func humanizeObjects(n int, d int) string {
 	return fmt.Sprintf("%.2f/%.2fM", float64(n)/m, float64(d)/m)
 }
 
-func (r *runner) processGogitStatus(
-	ctx context.Context, statusChan <-chan plumbing.StatusUpdate) {
+func (r *runner) printJournalStatusUntilFlushed(
+	ctx context.Context, doneCh <-chan struct{}) {
+	rootNode, _, err := r.config.KBFSOps().GetOrCreateRootNode(
+		ctx, r.h, libkbfs.MasterBranch)
+	if err != nil {
+		r.log.CDebugf(ctx, "GetOrCreateRootNode error: %+v", err)
+		return
+	}
+
+	err = r.config.KBFSOps().SyncAll(ctx, rootNode.GetFolderBranch())
+	if err != nil {
+		r.log.CDebugf(ctx, "SyncAll error: %+v", err)
+		return
+	}
+
+	jServer, err := libkbfs.GetJournalServer(r.config)
+	if err != nil {
+		r.log.CDebugf(ctx, "No journal server: %+v", err)
+	}
+
+	r.printJournalStatus(
+		ctx, jServer, rootNode.GetFolderBranch().Tlf, doneCh)
+}
+
+func (r *runner) processGogitStatus(ctx context.Context,
+	statusChan <-chan plumbing.StatusUpdate, fsEvents <-chan libfs.FSEvent) {
 	currStage := plumbing.StatusUnknown
 	var startTime time.Time
 	lastByteCount := 0
 	cpuProf := ""
-	for update := range statusChan {
-		if update.Stage != currStage {
-			if currStage != plumbing.StatusUnknown {
-				memProf := fmt.Sprintf("mem.%d.prof", currStage)
-				elapsedStr := r.getElapsedStr(ctx, startTime, memProf, cpuProf)
-				r.errput.Write([]byte("done." + elapsedStr + "\n"))
+	donePrinted := true
+	for {
+		if fsEvents == nil && statusChan == nil {
+			// Both channels are closed.
+			break
+		}
+		select {
+		case update, ok := <-statusChan:
+			if !ok {
+				statusChan = nil
+				continue
 			}
-			if r.verbosity >= 4 {
-				cpuProf = filepath.Join(
-					os.TempDir(), fmt.Sprintf("cpu.%d.prof", update.Stage))
-				f, err := os.Create(cpuProf)
-				if err != nil {
-					r.log.CDebugf(
-						ctx, "Couldn't create CPU profile: %s", cpuProf)
-					cpuProf = ""
-				} else {
-					pprof.StartCPUProfile(f)
+			if update.Stage != currStage {
+				if currStage != plumbing.StatusUnknown {
+					memProf := fmt.Sprintf("mem.%d.prof", currStage)
+					elapsedStr := r.getElapsedStr(ctx, startTime, memProf, cpuProf)
+					// go-git seems to grab the lock during this phase. So we'd
```

Comment:
It actually grabs the lock _after_ this phase, but before it sends the Done status update.

Do you need to hard-code the phase?  It seems like just `!donePrinted` would be enough, and more future-proof.

Reply:
Hmm good point. `donePrinted` was added later and I didn't realize it could replace the hard coded status. I'll update that as well as comment. Thanks!